spacy/visualization.py

```diff
@@ -0,0 +1,601 @@
+        whole_row_fg_color:                 a foreground color used for the whole row. This takes precedence over value_dependent_fg_colors.
+        whole_row_bg_color:                 a background color used for the whole row. This takes precedence over value_dependent_fg_colors.
+        """
+        obj = token
+        parts = self.attribute.split(".")
+        for part in parts[:-1]:
+            obj = getattr(obj, part)
+        value = str(getattr(obj, parts[-1]))
+        if self.permitted_values is not None and value not in (
+            str(v) for v in self.permitted_values
+        ):
+            return ""
+        if self.max_width is not None:
+            value = value[: self.max_width]
+        fg_color = None
+        bg_color = None
+        if right_pad_to_length is not None:
+            right_padding = " " * (right_pad_to_length - len(value))
+        else:
+            right_padding = ""
+        if not ignore_colors and len(value) > 0:
+            if whole_row_fg_color is not None:
+                fg_color = whole_row_fg_color
+            elif self.value_dependent_fg_colors is not None:
+                fg_color = self.value_dependent_fg_colors.get(value, None)
+            if fg_color is None and render_all_colors_within_values:
+                fg_color = self.fg_color
+            if self.value_dependent_bg_colors is not None:
+                bg_color = self.value_dependent_bg_colors.get(value, None)
+            if whole_row_bg_color is not None:
+                bg_color = whole_row_bg_color
+            elif bg_color is None and render_all_colors_within_values:
+                bg_color = self.bg_color
+        if fg_color is not None or bg_color is not None:
+            value = self.printer.text(value, color=fg_color, bg_color=bg_color)
+        return value + right_padding
+
+
+class Visualizer:
+    @staticmethod
+    def render_dependency_tree(sent: Span, root_right: bool) -> List[str]:
+        """
+        Returns an ASCII rendering of the document with a dependency tree for each sentence. The
+        dependency tree output for a given token has the same index within the output list of
+        strings as that token within the input document.
+
+        root_right: True if the tree should be rendered with the root on the right-hand side,
+                    False if the tree should be rendered with the root on the left-hand side.
+
+        Algorithm adapted from https://github.com/KoichiYasuoka/deplacy
+        """
+
+        # Check sent is really a sentence
+        assert sent.start == sent[0].sent.start
+        assert sent.end == sent[0].sent.end
+        heads: List[Optional[int]] = [
+            None
+            if token.dep_.lower() == "root" or token.head.i == token.i
+            else token.head.i - sent.start
+            for token in sent
+        ]
+        # Check there are no head references outside the sentence
```

Comment:
We should replace these naked asserts with actual explicit checks and raise a `ValueError` if needed. I think we can also refactor the conditional to make it more compact and take up fewer lines when auto-formatted.